src/v/s3/client.cc

```diff
@@ -0,0 +1,427 @@
+#include "s3/client.h"
+
+#include "bytes/iobuf.h"
+#include "bytes/iobuf_istreambuf.h"
+#include "hashing/secure.h"
+#include "s3/error.h"
+#include "s3/logger.h"
+#include "s3/signature.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/iostream.hh>
+#include <seastar/core/temporary_buffer.hh>
+#include <seastar/net/dns.hh>
+#include <seastar/net/inet_address.hh>
+#include <seastar/net/tls.hh>
+
+#include <boost/beast/core/error.hpp>
+#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/xml_parser.hpp>
+#include <fmt/core.h>
+#include <gnutls/crypto.h>
+
+namespace s3 {
+
+struct aws_header_names {
+    static constexpr boost::beast::string_view prefix = "prefix";
+    static constexpr boost::beast::string_view start_after = "start-after";
+    static constexpr boost::beast::string_view max_keys = "max-keys";
+    static constexpr boost::beast::string_view x_amz_content_sha256
+      = "x-amz-content-sha256";
+};
+
+struct aws_header_values {
+    static constexpr boost::beast::string_view user_agent
+      = "redpanda.vectorized.io";
+    static constexpr boost::beast::string_view text_plain = "text/plain";
+};
+
+// configuration //
+
+ss::future<configuration> configuration::make_configuration(
+  const public_key_str& pkey,
+  const private_key_str& skey,
+  const aws_region_name& region) {
+    return ss::do_with(
+      configuration(),
+      ss::tls::credentials_builder(),
+      [pkey, skey, region](
+        configuration& client_cfg, ss::tls::credentials_builder& cred_builder) {
+          const auto endpoint_uri = fmt::format(
+            "s3.{}.amazonaws.com", region());
+          // Setup credentials for TLS
+          ss::tls::credentials_builder builder;
+          client_cfg.access_key = pkey;
+          client_cfg.secret_key = skey;
+          client_cfg.region = region;
+          client_cfg.uri = access_point_uri(endpoint_uri);
+          // NOTE: this is a pre-defined gnutls priority string that
+          // picks the the ciphersuites with 128-bit ciphers which
+          // leads to up to 10x improvement in upload speed, compared
+          // to 256-bit ciphers
+          cred_builder.set_priority_string("PERFORMANCE");
+          return cred_builder.set_system_trust()
+            .then([&client_cfg, &cred_builder] {
+                return cred_builder.build_reloadable_certificate_credentials()
```

Comment:
in the future, this looks like a great candidate for a `coawait` transform and probably reduce line length by 2-3x. 

```
auto cfg = configuration();
...
coawait cred.set_system_trust();
coawait dns::resolve();
co_return client_cfg;

```